server/channels/db/migrations/postgres/000122_preferences_value_length.up.sql

```diff
@@ -0,0 +1 @@
+ALTER TABLE Preferences ALTER COLUMN TYPE text;
```

Comment:
Aren't we missing the column name in this statement (and in the .down.sql as well)?

Reply:
> Postgres hasn't required a table rewrite to do that since 9.2

Oh, this is interesting. From [the docs on ALTER TABLE](https://www.postgresql.org/docs/11/sql-altertable.html#SQL-ALTERTABLE-NOTES):
> changing the type of an existing column will require the entire table and its indexes to be rewritten. As an exception, when changing the type of an existing column, if the USING clause does not change the column contents and the old type is either binary coercible to the new type or an unconstrained domain over the new type, a table rewrite is not needed

This migration is not changing the column contents, and according to [this other docs page](https://www.postgresql.org/docs/current/sql-createcast.html), varchar and text are binary coercible:

> Two types can be binary coercible, which means that the conversion can be performed “for free” without invoking any function. This requires that corresponding values use the same internal representation. For instance, the types text and varchar are binary coercible both ways

So it seems that there is no need for a table rewrite here indeed.

Reply:
The question I still have is: are concurrent DML allowed? I haven't found an answer in the docs for that one.

Reply:
cc/@agnivade

Reply:
Regarding the type change, I found that from this SO response and the linked changelog line

https://stackoverflow.com/a/55294063

> Reduce need to rebuild tables and indexes for certain [ALTER TABLE](https://www.postgresql.org/docs/9.2/sql-altertable.html) ... ALTER COLUMN TYPE operations (Noah Misch)
>
> Increasing the length limit for a varchar or varbit column, or removing the limit altogether, no longer requires a table rewrite. Similarly, increasing the allowable precision of a numeric column, or changing a column from constrained numeric to unconstrained numeric, no longer requires a table rewrite. Table rewrites are also avoided in similar cases involving the interval, timestamp, and timestamptz types.

Reply:
This was a really interesting migration. And the devil is in the details.

> Adding a column with a volatile DEFAULT or changing the type of an existing column will require the entire table and its indexes to be rewritten. As an exception, when changing the type of an existing column, if the USING clause does not change the column contents and the old type is either binary coercible to the new type or an unconstrained domain over the new type, a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt.


So as you both have figured out, since TEXT and varchar columns are coercible, so therefore no table rewrites will happen.

BUT, look at the last sentence.

> ; but any indexes on the affected columns must still be rebuilt.

This is an important one. Because I tried the same change on the `hashtags` column on the posts table, and that took a long time. This is because even if the column does not need to be changed, the indexes will still need to be rebuilt. But we are lucky here because there are no indexes on the `value` column.

> The question I still have is: are concurrent DML allowed? I haven't found an answer in the docs for that one.

@agarciamontoro - Concurrent DML is not allowed during the course of the migration. As mentioned [here](https://www.postgresql.org/docs/current/sql-altertable.html#id-1.9.3.35.5)

>  An ACCESS EXCLUSIVE lock is acquired unless explicitly noted.

So a full lock on the table metadata will always happen. It's not possible to get rid of that. But that only takes a few miliseconds. So as long as there are no rewrites, we are fine. In this case, we got lucky because there's no index.

Reply:
Preferences is a pretty dumb table that the server doesn't know much about which certainly has some downsides (like how the client loads the user's preferences), but that does have upsides at least like how this sort of operation is easy 😀